libeos-update-server/server.c

```diff
@@ -0,0 +1,401 @@
+} EusServerProperty;
+
+static GParamSpec *props[PROP_LAST_REQUEST_TIME + 1] = { NULL, };
+
+static void request_read_cb (SoupServer        *soup_server,
+                             SoupMessage       *message,
+                             SoupClientContext *client,
+                             gpointer           user_data);
+static void request_finished_cb (SoupServer        *soup_server,
+                                 SoupMessage       *message,
+                                 SoupClientContext *client,
+                                 gpointer           user_data);
+static void request_aborted_cb (SoupServer        *soup_server,
+                                SoupMessage       *message,
+                                SoupClientContext *client,
+                                gpointer           user_data);
+
+static void
+eus_server_init (EusServer *self)
+{
+  self->repos = g_ptr_array_new_with_free_func (g_object_unref);
+}
+
+static void
+eus_server_constructed (GObject *object)
+{
+  EusServer *self = EUS_SERVER (object);
+
+  G_OBJECT_CLASS (eus_server_parent_class)->constructed (object);
+
+  g_signal_connect (self->server, "request-read", (GCallback) request_read_cb, self);
+  g_signal_connect (self->server, "request-finished", (GCallback) request_finished_cb, self);
+  g_signal_connect (self->server, "request-aborted", (GCallback) request_aborted_cb, self);
+}
+
+static void
+eus_server_get_property (GObject    *object,
+                         guint       property_id,
+                         GValue     *value,
+                         GParamSpec *spec)
+{
+  EusServer *self = EUS_SERVER (object);
+
+  switch ((EusServerProperty) property_id)
+    {
+    case PROP_SERVER:
+      g_value_set_object (value, self->server);
+      break;
+
+    case PROP_PENDING_REQUESTS:
+      g_value_set_uint (value, self->pending_requests);
+      break;
+
+    case PROP_LAST_REQUEST_TIME:
+      g_value_set_int64 (value, self->last_request_time);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, spec);
+      break;
+    }
+}
+
+static void
+eus_server_set_property (GObject      *object,
+                         guint         property_id,
+                         const GValue *value,
+                         GParamSpec   *spec)
+{
+  EusServer *self = EUS_SERVER (object);
+
+  switch ((EusServerProperty) property_id)
+    {
+    case PROP_SERVER:
+      g_set_object (&self->server, g_value_get_object (value));
+      break;
+
+    case PROP_PENDING_REQUESTS:
+    case PROP_LAST_REQUEST_TIME:
+      /* Read only. */
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, spec);
+      break;
+    }
+}
+
+static void
+eus_server_dispose (GObject *object)
+{
+  EusServer *self = EUS_SERVER (object);
+
+  self->pending_requests = 0;
+  self->last_request_time = 0;
+  g_clear_pointer (&self->repos, g_ptr_array_unref);
+
+  if (self->server != NULL)
```

Comment:
Not correct: `dispose()` can be called multiple times, and it must not crash the second time it’s called. (That’s really unlikely to happen, as it only needs to happen when getting rid of refcount loops, but it’s best to write all `dispose()` implementations robustly.)